pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/EquivalentStagesFinder.java

```diff
@@ -0,0 +1,339 @@
+import org.apache.pinot.query.planner.plannode.WindowNode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * This utility class can be used to find equivalent stages in the query plan.
+ *
+ * Equivalent stages are stages that represent the same job to be done. These stages can be potentially optimized to
+ * execute that job only once in a special stage that broadcast the results to all the equivalent stages.
+ */
+public class EquivalentStagesFinder {
+  public static final Logger LOGGER = LoggerFactory.getLogger(EquivalentStagesFinder.class);
+
+  private EquivalentStagesFinder() {
+  }
+
+  public static GroupedStages findEquivalentStages(MailboxSendNode root) {
+    Visitor visitor = new Visitor();
+    root.visit(visitor, null);
+
+    return visitor._equivalentStages;
+  }
+
+  /**
+   * A visitor that iterates the plan tree and finds equivalent stages.
+   *
+   * It may be a bit confusing that this class, which ends up being a visitor, calls another visitor to compare nodes.
+   * The reason is that this object implements visitor to iterate the plan tree in pre-order. Then for each
+   * mailbox send node (which are always the root of a stage), it calls
+   * {@link NodeEquivalence#areEquivalent(MailboxSendNode, MailboxSendNode)}. NodeEquivalence is another class that
+   * implements visitor, but this time to compare two nodes.
+   */
+  private static class Visitor extends PlanNodeVisitor.DepthFirstVisitor<Void, Void> {
+    private final GroupedStages.Mutable _equivalentStages = new GroupedStages.Mutable();
+    private final NodeEquivalence _nodeEquivalence = new NodeEquivalence();
+
+    @Override
+    public Void visitMailboxSend(MailboxSendNode node, Void context) {
+      // It is important to visit children before doing anything.
+      // This is a requirement on NodeEquivalence.areEquivalent() method that reduce the complexity of the algorithm
+      // from O(n^3) to O(n^2).
+      visitChildren(node, context);
+
+      // first try to find if the current node/stage is equivalent to an already equivalent stages.
+      for (MailboxSendNode uniqueStage : _equivalentStages.getLeaders()) {
+        if (_nodeEquivalence.areEquivalent(node, uniqueStage)) {
+          _equivalentStages.addToGroup(uniqueStage, node);
+          return null;
+        }
+      }
+      // there is no visited stage that is equivalent to the current stage, so add it to the unique visited stages.
+      _equivalentStages.addNewGroup(node);
+      return null;
+    }
+
+    /**
+     * A visitor that compares two nodes to see if they are equivalent.
+     *
+     * The implementation uses the already visited stages (stored in {@link #_equivalentStages}) to avoid comparing the
+     * same nodes multiple times. The side effect of that is that the second argument for {@link #areEquivalent} must be
+     * a node that was already visited.
+     */
+    private class NodeEquivalence implements PlanNodeVisitor<Boolean, PlanNode> {
```

Comment:
I need to step through patch to understand. Where is the return value (Boolean) checked ? 
Also when are methods like `visitAggregate` called as  I cannot find `visitChildren` calls in the `NodeEquivalence` class.

Reply:
> Where is the return value (Boolean) checked ?

The entry point here is `areEquivalent`. That method returns the `Boolean` returned by each `visit` method. And `areEquivalent` is being called by `Visitor` in its `visitMailboxSend`.

> Also when are methods like visitAggregate called as I cannot find visitChildren calls in the NodeEquivalence class.

There is no `visitChildren` in this method. `NodeEquivalence` is a normal visitor, not a `DepthFirstVisitor`. Recursion is done in baseNode, which is called by all visit specific methods.